### DISCUSSION TITLE
Elasticsearch Intergration

### DISCUSSION THREAD

**Discussion by imiskolee** [directus/directus#12264] [Feature Requests]
# Background

when we searched on a big table(over 10M records), it was very slow now. because SQL like search can not hit any index, integrate ElasticSearch replace SQL like is a very general solution for full text search. 

# Sub Tasks

* add  `ElasticSearch Support` flag on collection configure panel.
* automatically mapping & sync data by database hook.
*  route to elasticsearch.

**Comment by rijkvanzanten** [comment]
Love it! I've been thinking through making "search adapters" another extension type. That way, you could replace the default "local sql" search with another adapter, like elasticsearch, algolia, or any other third party tool 🙂

**Reply by imiskolee** [reply]
> 

I am thinking two ways:

#1. use API Hook to 100% replace default logics
#2. let Elasticsearch as a flag.

**Reply by rijkvanzanten** [reply]
> https://github.com/directus/directus/issues/2. let Elasticsearch as a flag.

That'd mean that we'd have to hardcode support for just elasticsearch. It will get the job done, but it will be 1 day before somebody shows up to ask to support <insert search engine> as well 😅

**Reply by bamboowonder** [reply]
+1 on this.  Here is a link to my  related question on discord:  https://discord.com/channels/725371605378924594/741330910594465823/961717429573128212

Postgres with good indexing (fulltext or trigram) has great performance for searching across a large table.

All we'd need to do is be able to modify the WHERE clause to use: ftsvectorcolumn @@ websearch_to_tsquery('chocolate dark')   instead of ILIKE or ANY or IN.

We'd be interested in (partially) sponsoring a feature like this.

**Reply by imiskolee** [reply]
> +1 on this. Here is a link to my related question on discord: https://discord.com/channels/725371605378924594/741330910594465823/961717429573128212
> 
> Postgres with good indexing (fulltext or trigram) has great performance for searching across a large table.
> 
> All we'd need to do is be able to modify the WHERE clause to use: ftsvectorcolumn @@ websearch_to_tsquery('chocolate dark') instead of ILIKE or ANY or IN.
> 
> We'd be interested in (partially) sponsoring a feature like this.

yes. MySQL also has full-text index feature, another option is to develop a plugin(hook) to full control search API, somethings like:

```
//Search API

const hook = this.overwideHooks["query"]

if(hook) {
  let response = hook(request_context)
  return response
}

//generic logic here...


```

**Comment by imiskolee** [comment]
add another request: https://github.com/directus/directus/discussions/12250

**Comment by diegoleme** [comment]
@rijkvanzanten I really like the idea `search adapters`.

What would need to be done here for this issue to gain priority? Can I help in any way?

**Reply by rijkvanzanten** [reply]
Either the feature gets a lot of upvotes/traction from the community to indicate that this is something worth doing early or a party can come along and sponsor it's development 🙂 

In any case, the first thing we should do is put together a little RFC on how this should ideally work. I'd imagine a new environment variable that effectively controls where `search` is being handled. I'm curious to see if we can find some sort of way to properly "sync" with these external services, as I believe that some/all/most will have to have some binding to the database to index the contents?

**Reply by diegoleme** [reply]
I think there are 3 main steps:

1. Configure where the search request goes;
2. Standardize a format for how layouts will parse the search response;
3. How to synchronize the databases;

**Reply by rijkvanzanten** [reply]
I think number 2 should be standardized API side, rather than App side. The output of the API should be identical regardless of search backend provider used

**Reply by imiskolee** [reply]
the step 3 is a ETL task, i think user can use third-party for this.

**Reply by diegoleme** [reply]
Following these thoughts, we will only have to solve point 1, configuring a URL to do the search.

Another way would be to integrate somehow with some standard search tool, and make it a directus feature

**Comment by cdwmhcc** [comment]
I've been using Elasticsearch a long time ago to sync data to es via hooks.I think the priority of integration should be improved. Search is a very basic function, but if you want accurate search, it is a function that relies heavily on algorithms.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by AdolfVonKleist** [reply]
Is there any further update on this?

**Comment by JayShoe** [comment]
Instead of trying to force extendable search engines into the current collection, filter, or field parameters, I think we should look at adding a more flexible interface. This would let us define and integrate other search concepts more easily without overcomplicating the existing setup.

My proposal for facets could help with this. It introduces a way to handle search filters in a flexible way that could support new search concepts while keeping things consistent for parsing and syncing databases.

Here’s the link to the discussion: https://github.com/directus/directus/discussions/23605. Let me know what you think!

**Reply by rijkvanzanten** [reply]
Good shout! Especially with the rise of things like vector search the "standard" of what the various searching parameters look like is changing rapidly
